// @gnosis.pm/safe-apps-sdk/dist/index.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
            if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                t[p[i]] = s[p[i]];
        }
    return t;
};
var __spreadArrays = (this && this.__spreadArrays) || function () {
    for (var s = 0, i = 0, il = arguments.length; i < il; i++) s += arguments[i].length;
    for (var r = Array(s), k = 0, i = 0; i < il; i++)
        for (var a = arguments[i], j = 0, jl = a.length; j < jl; j++, k++)
            r[k] = a[j];
    return r;
};
Object.defineProperty(exports, "__esModule", { value: true });
var FromSafeMessages;
(function (FromSafeMessages) {
    FromSafeMessages["ON_SAFE_INFO"] = "ON_SAFE_INFO";
})(FromSafeMessages = exports.FromSafeMessages || (exports.FromSafeMessages = {}));
var ToSafeMessages;
(function (ToSafeMessages) {
    ToSafeMessages["SEND_TRANSACTIONS"] = "SEND_TRANSACTIONS";
})(ToSafeMessages = exports.ToSafeMessages || (exports.ToSafeMessages = {}));
var config = {};
var _logMessageFromSafe = function (origin, message) {
    console.info("SafeConnector: A message with id " + message + " was received from origin " + origin + ".");
};
var _onParentMessage = function (_a) {
    var origin = _a.origin, data = _a.data;
    return __awaiter(void 0, void 0, void 0, function () {
        var _b;
        return __generator(this, function (_c) {
            if (origin === window.origin) {
                return [2 /*return*/];
            }
            if (((_b = config.safeAppUrlsRegExp) === null || _b === void 0 ? void 0 : _b.find(function (regExp) { return regExp.test(origin); })) === undefined) {
                console.error("SafeConnector: A message was received from an unknown origin: " + origin + ".");
                return [2 /*return*/];
            }
            if (!data || !data.messageId) {
                console.error("SafeConnector: A message was received from origin " + origin + " with NO message id provided.");
                return [2 /*return*/];
            }
            if (!config.listeners) {
                console.error("SafeConnector: A message was received from origin " + origin + " but no listeners were registered.");
                return [2 /*return*/];
            }
            switch (data.messageId) {
                case FromSafeMessages.ON_SAFE_INFO: {
                    _logMessageFromSafe(origin, FromSafeMessages.ON_SAFE_INFO);
                    console.log(data.data);
                    config.listeners.onSafeInfo({
                        safeAddress: data.data.safeAddress,
                        network: data.data.network.toLowerCase(),
                        ethBalance: data.data.ethBalance,
                    });
                    break;
                }
                default: {
                    console.warn("SafeConnector: A message was received from origin " + origin + " with an unknown message id: " + data.messageId);
                    break;
                }
            }
            return [2 /*return*/];
        });
    });
};
var _sendMessageToParent = function (messageId, data) {
    window.parent.postMessage({ messageId: messageId, data: data }, '*');
};
/**
 * Register all the listeners supported. When Safe-app sends a message
 * depending on the messageId, the corresponding listener will be called
 * @param listeners
 */
function addListeners(_a) {
    var allListeners = __rest(_a, []);
    config.listeners = __assign({}, allListeners);
    window.addEventListener('message', _onParentMessage);
}
/**
 * Unregister all the listeners previously set by addListeners.
 */
function removeListeners() {
    window.removeEventListener('message', _onParentMessage);
}
/**
 * Request Safe app to send transactions
 * @param txs
 */
function sendTransactions(txs) {
    if (!txs || !txs.length) {
        return;
    }
    _sendMessageToParent(ToSafeMessages.SEND_TRANSACTIONS, txs);
}
/**
 * Sets Safe-app url that will render the third-party app.
 * @param parentUrl
 */
function initSdk(safeAppUrlsRegExp) {
    if (safeAppUrlsRegExp === void 0) { safeAppUrlsRegExp = []; }
    config.safeAppUrlsRegExp = __spreadArrays([
        /https:\/\/.*(gnosis-safe\.io|gnosisdev.com)/,
        /https?:\/\/localhost:\d+/
    ], safeAppUrlsRegExp);
    return { addListeners: addListeners, removeListeners: removeListeners, sendTransactions: sendTransactions };
}
exports.default = initSdk;
